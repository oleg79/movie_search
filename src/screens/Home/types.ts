import {Video} from '../../components/VideoCard';

export interface SuccessResponse {
  Response: string;
  Search: Video[];
  totalResults: string;
}

export interface FailResponse {
  Response: string;
  Error: string;
}

export type ApiResponse = SuccessResponse | FailResponse | null;