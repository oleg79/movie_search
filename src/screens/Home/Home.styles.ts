import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  root: {
    flex: 1,
  },

  search: {
    width: '90%',
    alignSelf: 'center',
    fontSize: 20,
    paddingVertical: 10,
  },

  searchButton: {
    width: '90%',
    textAlign: 'center',
    alignSelf: 'center',
    paddingVertical: 12,
    backgroundColor: '#e53935',
    color: '#fff',
    fontSize: 15,
    fontWeight: 'bold',
  },

  resultInfoContainer: {
    width: '90%',
    alignSelf: 'center',
  },

  resultInfo: {
    textAlign: 'center',
    paddingVertical: 10,
    color: '#37474f',
  },
});
