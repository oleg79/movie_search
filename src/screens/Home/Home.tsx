import * as React from 'react';
import {
  Text,
  TextInput,
  TouchableOpacity,
  View
} from 'react-native';
import Config from 'react-native-config'

import styles from './Home.styles'

interface SuccessResponse {
  Response: string;
  Search: Object[];
  totalResults: string;
}

interface FailResponse {
  Response: string;
  Error: string;
}

type Response = SuccessResponse | FailResponse | null;

interface State {
  searchText: string;
  showEmptySearchWarning: boolean;
  searching: boolean;
  response: Response;
}

export default class Home extends React.PureComponent<{}, State> {
  state = {
    searchText: '',
    showEmptySearchWarning: false,
    searching: false,
    response: null,
  };

  handleSearchTextChange: (searchText: string) => void = searchText => {
    this.setState({
      searchText,
      showEmptySearchWarning: false
    });
  };

  handleSearchPress: () => void = () => {
    const searchText = this.state.searchText.trim();

    if (searchText) {
      this.search(searchText);
    } else {
      this.setState({
        showEmptySearchWarning: true
      });
    }
  };

  search: (searchText: string) => void = async searchText => {
    const rawResponse = await fetch(`${Config.API_URL}/?apikey=${Config.API_KEY}&s=${searchText}`);
    const response = await rawResponse.json();
    this.setState({ response })
  };

  render() {
    return (
      <View style={styles.root}>
        <TextInput
          placeholder='Enter movie title'
          onChangeText={this.handleSearchTextChange}
        />
        <TouchableOpacity onPress={this.handleSearchPress}>
          <Text>Search</Text>
        </TouchableOpacity>

        { this.state.showEmptySearchWarning &&
          <Text>Please enter the title</Text>
        }
      </View>
    );
  }
}
