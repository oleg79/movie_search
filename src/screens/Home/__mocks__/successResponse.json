{
  "Response": "True",
  "Search": [
    {
      "Poster": "https://m.media-amazon.com/images/M/MV5BMzFkM2YwOTQtYzk2Mi00N2VlLWE3NTItN2YwNDg1YmY0ZDNmXkEyXkFqcGdeQXVyMTMxODk2OTU@._V1_SX300.jpg",
      "Title": "Home Alone",
      "Type": "movie",
      "Year": "1990",
      "imdbID": "tt0099785"
    }
  ],
  "totalResults": "4184"
}