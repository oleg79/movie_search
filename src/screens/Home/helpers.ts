import {
  FailResponse,
  SuccessResponse,
  ApiResponse
} from "./types";

export const isSuccessResponse = (response: ApiResponse): response is SuccessResponse => {
  if (response === null) {
    return false;
  }
  return (<SuccessResponse>response).Search !== undefined;
};

export const isFailResponse = (response: ApiResponse): response is FailResponse => {
  if (response === null) {
    return false;
  }
  return (<FailResponse>response).Error !== undefined;
};