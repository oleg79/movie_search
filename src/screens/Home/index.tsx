import * as React from 'react';
import {
  ActivityIndicator,
  FlatList,
  Text,
  TextInput,
  TouchableOpacity,
  View
} from 'react-native';
import Config from 'react-native-config';
import {NavigationInjectedProps} from 'react-navigation';

import axios from 'axios';
import { SearchParamsContext } from '../../providers/SearchParamsProvider';

import {Video, VideoCard} from '../../components/VideoCard';

import { isSuccessResponse, isFailResponse } from './helpers';
import styles from './Home.styles'

interface State {
  page: number;
  searchText: string;
  showEmptySearchWarning: boolean;
  loading: boolean;
  loadingMore: boolean;
  data: Video[];
  error: string | null;
  totalResults: number;
}

export const defaultState: State = {
  page: 1,
  searchText: '',
  showEmptySearchWarning: false,
  loading: false,
  loadingMore: false,
  data: [],
  error: null,
  totalResults: 0,
};

export default class Home extends React.PureComponent<NavigationInjectedProps, State> {
  static contextType = SearchParamsContext;
  context!: React.ContextType<typeof SearchParamsContext>;

  offset: number = 0;

  state = defaultState;

  handleSearchTextChange: (searchText: string) => void = searchText => {
    this.setState({
      searchText,
      showEmptySearchWarning: false,
      error: null,
    });
  };

  handleSearchPress: () => void = () => {
    const searchText = this.state.searchText.trim();

    if (searchText) {
      this.setState({
        page: 1,
        data: [],
        loading: true,
      }, () => {
        this.search();
      });

    } else {
      this.setState({
        showEmptySearchWarning: true
      });
    }
  };

  search: () => void = async () => {
    const { page, searchText } = this.state;
    const { year, type, isActive } = this.context;

    let uri = `${Config.API_URL}/?apikey=${Config.API_KEY}&s=${searchText.trim()}&page=${page}`;

    if (isActive) {
      uri += `&y=${year}${type ? '&type=' + type : ''}`;
    }

    try {
      const { data: response } = await axios.get(uri);
      if (isSuccessResponse(response)) {
        this.setState(() => ({
          data: page === 1 ? response.Search : [ ...this.state.data, ...response.Search ],
          totalResults: parseInt(response.totalResults, 10),
          loading: false,
          loadingMore: false,
          error: null,
        }));
      } else if (isFailResponse(response)) {
        this.setState({
          error: response.Error,
          totalResults: 0,
          loading: false,
          loadingMore: false,
        });
      }
    } catch (e) {
      this.setState({ error: e.message });
    }

  };

  renderItem = ({item}:{item: Video}) => <VideoCard video={item} navigation={this.props.navigation}/>;

  keyExtractor: (item: Video) => string = (item: Video) => item.imdbID;

  loadMoreVideos: () => void = () => {
    if (this.state.totalResults !== this.state.data.length) {
      this.setState((prevState) => ({
        page: prevState.page + 1,
        loadingMore: true,
      }), () => {
        this.search();
      });
    }
  };

  renderListFooter = () => {
    if (!this.state.loadingMore) {
      return null;
    }

    return (
      <View>
        <ActivityIndicator animating size='large' />
      </View>
    );
  };

  render() {
    const {
      totalResults,
      data,
      showEmptySearchWarning,
      error,
    } = this.state;

    return (
      <View style={styles.root}>
        <View>
          <TextInput
            placeholder='Enter movie title'
            onChangeText={this.handleSearchTextChange}
            value={this.state.searchText}
            style={styles.search}
          />
          <TouchableOpacity onPress={this.handleSearchPress}>
            <Text style={styles.searchButton}>Search</Text>
          </TouchableOpacity>
        </View>


        { Boolean(totalResults) &&
          <View style={styles.resultInfoContainer}>
            <Text style={styles.resultInfo}>
              {data.length} results of { totalResults } are loaded
            </Text>
          </View>
        }

        { showEmptySearchWarning &&
          <View style={styles.errorContainer}>
            <Text style={styles.errorText}>Please enter a title</Text>
          </View>
        }

        {!error &&
          <FlatList
            data={data}
            renderItem={this.renderItem}
            keyExtractor={this.keyExtractor}
            initialNumToRender={3}
            onEndReachedThreshold={0.5}
            onEndReached={this.loadMoreVideos}
            ListFooterComponent={this.renderListFooter}
          />
        }

        { error &&
          <View style={styles.errorContainer}>
            <Text style={styles.errorText}>{ error }</Text>
          </View>
        }
      </View>
    );
  }
}
