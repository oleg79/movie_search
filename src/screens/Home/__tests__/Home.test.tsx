import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import successResponse from '../__mocks__/successResponse.json';
import failResponse from '../__mocks__/failResponse.json';

import Home from '../Home';

const generateFetch = (rawJsonResponse: string) => {
  // @ts-ignore
  global.fetch = jest.fn().mockImplementation(() => {
    return new Promise((resolve, reject) => {
      resolve({
        ok: true,
        Id: '123',
        json: function() {
          return Promise.resolve(JSON.parse(rawJsonResponse));
        }
      });
    });
  });
};

describe('<Home/> component', () => {
  let rendered: any;
  let instance: any;

  beforeEach(() => {
    rendered = renderer.create(<Home/>);
    instance = rendered.getInstance();
  });


  it('Should be rendered properly', () => {
    const tree = rendered.toJSON();

    expect(tree).toMatchSnapshot();
  });


  it('Should have a proper initial state', () => {
    expect(instance.state).toMatchObject({
      searchText: '',
      showEmptySearchWarning: false,
      searching: false,
      response: null,
    });
  });


  it('Should handle text change', () => {
    instance.handleSearchTextChange('moon');
    expect(instance.state.searchText).toBe('moon');
  });


  it('Should handle empty keyword search properly', () => {
    instance.handleSearchPress();
    expect(instance.state.showEmptySearchWarning).toBe(true);

    instance.handleSearchTextChange('    ');
    expect(instance.state.showEmptySearchWarning).toBe(false);

    instance.handleSearchPress();
    expect(instance.state.showEmptySearchWarning).toBe(true);
  });


  it('Should handle successful keyword search properly', async () => {
    generateFetch(JSON.stringify(successResponse));
    instance.handleSearchTextChange('some text');
    await instance.search('some text');
    expect(instance.state.response).toMatchObject(successResponse);
  });


  it('Should handle failed keyword search properly', async () => {
    generateFetch(JSON.stringify(failResponse));
    instance.handleSearchTextChange('some text');
    await instance.search('some text');
    expect(instance.state.response).toMatchObject(failResponse);
  });

});
