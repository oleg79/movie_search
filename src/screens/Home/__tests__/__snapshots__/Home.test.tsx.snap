// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home/> component Should be rendered properly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 40,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Enter movie title"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
  />
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      Search
    </Text>
  </View>
</View>
`;
